org.eclipse.jdt.core.tests.model/src/org/eclipse/jdt/core/tests/dom/CompilationUnitResolverDiscoveryTest.java

```diff
@@ -0,0 +1,184 @@
+			CompilationUnit compilationUnit = (CompilationUnit) result;
+			ASTNode node = getASTNode(compilationUnit, 0, 0);
+			assertEquals("Not a compilation unit", ASTNode.METHOD_DECLARATION, node.getNodeType());
+			MethodDeclaration methodDeclaration = (MethodDeclaration) node;
+			IMethodBinding mb = methodDeclaration.resolveBinding();
+			assertEquals(mb.getClass().getName(), "org.eclipse.jdt.core.dom.MethodBinding");
+		} finally {
+			System.setProperty(SELECTED_SYSPROP, original);
+		}
+	}
+
+	public void testCompilationUnitResolverValidSysprop() throws JavaModelException {
+		String SELECTED_SYSPROP = "ICompilationUnitResolver";
+		String original = System.getProperty(SELECTED_SYSPROP);
+		try {
+			System.setProperty(SELECTED_SYSPROP, "org.eclipse.jdt.core.tests.model.resolver1");
+			ICompilationUnit sourceUnit = getCompilationUnit("Converter9" , "src", "testBug497719_001", "X.java"); //$NON-NLS-1$ //$NON-NLS-2$ //$NON-NLS-3$ //$NON-NLS-4$
+			ASTNode result = runConversion(this.ast.apiLevel(), sourceUnit, true, true);
+			assertTrue("Not a compilation unit", result.getNodeType() == ASTNode.COMPILATION_UNIT);
+			CompilationUnit compilationUnit = (CompilationUnit) result;
+			PackageDeclaration pd = compilationUnit.getPackage();
+			assertNotNull(pd);
+			String pdName = pd.getName().toString();
+			assertEquals(pdName, "compilationUnitResolverDiscoveryTest");
+		} finally {
+			System.setProperty(SELECTED_SYSPROP, original);
+		}
+	}
+
+
+	/*
+	 * Custom made interface that implements exactly to the test
+	 */
+	public static final class TEST_RESOLVER implements ICompilationUnitResolver {
+		@Override
+		public CompilationUnit toCompilationUnit(org.eclipse.jdt.internal.compiler.env.ICompilationUnit sourceUnit,
+				boolean initialNeedsToResolveBinding, IJavaProject project, List<Classpath> classpaths,
+				int focalPosition, int apiLevel, Map<String, String> compilerOptions,
+				WorkingCopyOwner parsedUnitWorkingCopyOwner, WorkingCopyOwner typeRootWorkingCopyOwner, int flags,
+				IProgressMonitor monitor) {
+			// Return a mostly-invalid hard-coded dom tree for the purpose of this test
+			AST ast = AST.newAST(apiLevel, JavaCore.ENABLED.equals(compilerOptions.get(JavaCore.COMPILER_PB_ENABLE_PREVIEW_FEATURES)));
+			CompilationUnit res = ast.newCompilationUnit();
+			PackageDeclaration pack = ast.newPackageDeclaration();
+			pack.setName(ast.newSimpleName("compilationUnitResolverDiscoveryTest"));
+			res.setPackage(pack);
+			return res;
+		}
+
+		@Override
+		public void resolve(String[] sourceFilePaths, String[] encodings, String[] bindingKeys,
+				FileASTRequestor requestor, int apiLevel, Map<String, String> compilerOptions,
+				List<Classpath> classpathList, int flags, IProgressMonitor monitor) {
+		}
+
+		@Override
+		public void parse(ICompilationUnit[] compilationUnits, ASTRequestor requestor, int apiLevel,
+				Map<String, String> compilerOptions, int flags, IProgressMonitor monitor) {
+			// TODO Auto-generated method stub
```

Comment:
open TODOs -> mark as draft?

Reply:
As it's just testing the extension point mechanism, the implementation doesn't matter. This TODO should be replaced with some `// does not matter for test`